### PR TITLE
WI00713829 - Ports dont update when resizing nodes in Blazor Diagrams

### PR DESCRIPTION
**Pull request by Alvarenga1**
Fixed an issue causing port position to not be calculated during resizing of nodes.

### CODE REVIEW

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs [hunk 1]
```diff
@@ -166,6 +171,47 @@ private void UpdatePortPositions(double deltaX, double deltaY)
...
+                    port.Position = new Point(port.Position.X + deltaX / 2, port.Position.Y + deltaY);
+                    break;
+                case PortAlignment.BottomRight:
+                    port.Position = new Point(port.Position.X + deltaX, port.Position.Y + deltaY);
+                    break;
+                case PortAlignment.BottomLeft:
+                    port.Position = new Point(port.Position.X, port.Position.Y + deltaY);
+                    break;
+            }
+            port.RefreshLinks();
```
Comment (Heathermcx) [inline src/Blazor.Diagrams.Core/Models/NodeModel.cs:211]: Does this logic belong in the PortModel?

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs [hunk 2]
```diff
@@ -166,6 +171,47 @@ private void UpdatePortPositions(double deltaX, double deltaY)
         }
     }
 
+    /// <summary>
+    /// Updates the position of the ports by calculating the difference between the old and new size when the node is resized
```
Comment (Heathermcx) [inline src/Blazor.Diagrams.Core/Models/NodeModel.cs:175]:
```suggestion
    /// Updates port positions when node size changes.
```

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs [hunk 3]
```diff
@@ -166,6 +171,47 @@ private void UpdatePortPositions(double deltaX, double deltaY)
...
+    /// </summary>
+    private void UpdatePortPositions(double oldWidth, double oldHeight, double newWidth, double newHeight)
```
Comment (Heathermcx) [inline src/Blazor.Diagrams.Core/Models/NodeModel.cs:177]: maybe just pass in `Size oldSize` and `Size newSize`? I think that would be easier to read

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs [hunk 4]
```diff
@@ -156,6 +158,9 @@ public virtual void UpdatePositionSilently(double deltaX, double deltaY)
 
     public virtual bool CanAttachTo(ILinkable other) => other is not PortModel && other is not BaseLinkModel;
 
+    /// <summary>
+    /// Updates port position based on the deltaX and deltaY when the node is moved
```
Comment (Heathermcx) [inline src/Blazor.Diagrams.Core/Models/NodeModel.cs:162]:
```suggestion
    /// Updates port positions when node position changes.
```

**Review by Heathermcx — CHANGES_REQUESTED** [review]

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs (hunk 2 above)
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/NodeModel.cs:175]: Fixed

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs (hunk 3 above)
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/NodeModel.cs:177]: Fixed

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs (hunk 4 above)
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/NodeModel.cs:162]: Fixed

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs (hunk 1 above)
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/NodeModel.cs]: Fixed

**Review by Heathermcx — CHANGES_REQUESTED** [review]

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs [hunk 5]
```diff
@@ -166,6 +171,21 @@ private void UpdatePortPositions(double deltaX, double deltaY)
         }
     }
 
+    /// <summary>
+    /// Updates port positions when node size changes.
+    /// </summary>
+    private void UpdatePortPositions(Size oldSize, Size newSize)
+    {
+        var deltaX = newSize.Width - oldSize.Width;
+        var deltaY = newSize.Height - oldSize.Height;
```
[the remark this answers is not in the document]
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/NodeModel.cs:180]: I don't think so, they are used in the same way in both contexts, which is how much we need to offset the current port position by.

File: src/Blazor.Diagrams.Core/Models/PortModel.cs [hunk 6]
```diff
@@ -67,4 +67,38 @@ public virtual bool CanAttachTo(ILinkable other)
     void ILinkable.AddLink(BaseLinkModel link) => _links.Add(link);
 
     void ILinkable.RemoveLink(BaseLinkModel link) => _links.Remove(link);
+
+    public void calculateNewPortPosition(double deltaX, double deltaY)
```
[the remark this answers is not in the document]
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/PortModel.cs:71]: Fixed

File: tests/Blazor.Diagrams.Core.Tests/Models/NodeModelTest.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+            var newX = 200;
+            var newY = 300;
+
+            //Act
+            node.SetPosition(newX, newY);
+
+            //Assert
+            Assert.Equal(200, port.Position.X);
+            Assert.Equal(300, port.Position.Y);
```
[the remark this answers is not in the document]
Comment (Alvarenga1) [reply tests/Blazor.Diagrams.Core.Tests/Models/NodeModelTest.cs:36]: No, we're using the deltaX and deltaY to calculate the resulting port position by adding those values tothe current port position, this calculation is done in the UpdatePortPositions method in NodeModel.cs

File: src/Blazor.Diagrams.Core/Models/PortModel.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/PortModel.cs]: fixed

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs [hunk 8]
```diff
@@ -104,10 +104,12 @@ public override void SetPosition(double x, double y)
     public void SetSize(double width, double height)
     {
         var newSize = new Size(width, height);
-        if (newSize.Equals(_size) == true)
+        if (newSize.Equals(_size) == true || Size == null)
```
Comment (Heathermcx) [inline src/Blazor.Diagrams.Core/Models/NodeModel.cs:107]: I don't think we should return early if Size is null. We should still set the size. 
Maybe what you're looking for is something like 
```
if (Size != null) 
{
   UpdatePortPositions(Size, newSize);
}
Size = newSize;
```

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Heathermcx) [reply src/Blazor.Diagrams.Core/Models/NodeModel.cs:180]: I don't agree. They are used quite differently. If they were the same, we wouldn't need two `UpdatePortPositions` methods. In general we are refering to the position coordinates by `X` and `Y` and dimensions by `width` and `height` so I'm not sure why when we are talking about a change in position or dimensions we suddenly start using the same words to refer to different things.
Doesn't need to be `deltaHeight` and `deltaWidth` but I do think it should be different to `deltaX` and `deltaY` to improve the readability of the two methods.

File: tests/Blazor.Diagrams.Core.Tests/Blazor.Diagrams.Core.Tests.csproj
Comment (Heathermcx) [inline tests/Blazor.Diagrams.Core.Tests/Blazor.Diagrams.Core.Tests.csproj:1]: Do we need these changes?

File: tests/Blazor.Diagrams.Core.Tests/Models/NodeModelTest.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
+﻿using Blazor.Diagrams.Core.Geometry;
+using Blazor.Diagrams.Core.Models;
+using Moq;
+using Xunit;
+
+
```
Comment (Heathermcx) [inline tests/Blazor.Diagrams.Core.Tests/Models/NodeModelTest.cs:6]:
```suggestion

```

File: tests/Blazor.Diagrams.Core.Tests/Models/PortModelTest.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                    Assert.Equal(100, port.Position.Y);
+                    break;
+                case PortAlignment.BottomRight:
+                    Assert.Equal(100, port.Position.X);
+                    Assert.Equal(100, port.Position.Y);
+                    break;
+                case PortAlignment.BottomLeft:
+                    Assert.Equal(0, port.Position.X);
+                    Assert.Equal(100, port.Position.Y);
+                    break;
```
Comment (Heathermcx) [inline tests/Blazor.Diagrams.Core.Tests/Models/PortModelTest.cs:62]: I think that the expected values should be part of the `InlineData`, not as logic in the test.

**Review by Heathermcx — CHANGES_REQUESTED** [review]

File: tests/Blazor.Diagrams.Core.Tests/Models/NodeModelTest.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Heathermcx) [reply tests/Blazor.Diagrams.Core.Tests/Models/NodeModelTest.cs]: As discussed, the port position was not being initialized as expected which is why all of the ports were in the top left corner of the node after the position change. We have updated the test to initialize the port position and check that the post position is correctly set after the position change.

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/NodeModel.cs:180]: Fixed

File: src/Blazor.Diagrams.Core/Models/NodeModel.cs (hunk 8 above)
Comment (Alvarenga1) [reply src/Blazor.Diagrams.Core/Models/NodeModel.cs:107]: I don't think it makes sense to move the port before the node is resized, what do you think of this alternative approach ?
 ```csharp
public void SetSize(double width, double height)
    {
        var newSize = new Size(width, height);
        if (newSize.Equals(_size) == true)
            return;

        Size? oldSize = Size != null ? new Size(Size.Width, Size.Height) : null;

        Size = newSize;
        if (oldSize != null)
        {
            UpdatePortPositions(oldSize, newSize);
        }
        Refresh();
        RefreshLinks();
        SizeChanging?.Invoke(this);
    }
```

File: tests/Blazor.Diagrams.Core.Tests/Blazor.Diagrams.Core.Tests.csproj
Comment (Alvarenga1) [reply tests/Blazor.Diagrams.Core.Tests/Blazor.Diagrams.Core.Tests.csproj:1]: Reverted

File: tests/Blazor.Diagrams.Core.Tests/Models/NodeModelTest.cs (hunk 9 above)
Comment (Alvarenga1) [reply tests/Blazor.Diagrams.Core.Tests/Models/NodeModelTest.cs:6]: fixed

File: tests/Blazor.Diagrams.Core.Tests/Models/PortModelTest.cs (hunk 10 above)
Comment (Alvarenga1) [reply tests/Blazor.Diagrams.Core.Tests/Models/PortModelTest.cs:62]: fixed

File: tests/Blazor.Diagrams.Core.Tests/Models/PortModelTest.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+                case PortAlignment.BottomRight:
+                    Assert.Equal(expectedXPosition, port.Position.X);
+                    Assert.Equal(expectedYPosition, port.Position.Y);
+                    break;
+                case PortAlignment.BottomLeft:
+                    Assert.Equal(expectedXPosition, port.Position.X);
+                    Assert.Equal(expectedYPosition, port.Position.Y);
+                    break;
+            }
```
Comment (Heathermcx) [inline tests/Blazor.Diagrams.Core.Tests/Models/PortModelTest.cs:63]: we don't need this switch if in every case we are doing the same thing .

File: src/Blazor.Diagrams.Core/Models/PortModel.cs [hunk 12]
```diff
@@ -67,4 +67,38 @@ public virtual bool CanAttachTo(ILinkable other)
     void ILinkable.AddLink(BaseLinkModel link) => _links.Add(link);
 
     void ILinkable.RemoveLink(BaseLinkModel link) => _links.Remove(link);
+
+    public virtual void SetPortPositionOnNodeSizeChanged(double deltaX, double deltaY)
```
Comment (Heathermcx) [inline src/Blazor.Diagrams.Core/Models/PortModel.cs:71]: `deltaWidth`, `deltaHeight` for consistency

**Review by Heathermcx — COMMENTED** [review]
Hi Renan, just a couple more minor changes from feedback not entirely addressed in previous review.

File: tests/Blazor.Diagrams.Core.Tests/Models/PortModelTest.cs (hunk 11 above)
Comment (Alvarenga1) [reply tests/Blazor.Diagrams.Core.Tests/Models/PortModelTest.cs:63]: fixed

**Review by Heathermcx — APPROVED** [review]

**Pull request merged**